// rclnodejs/generated/robot_localization/robot_localization__srv__GetState_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const GetState_ResponseRefStruct = StructType({
  state: ArrayType(primitiveTypes.float64, 15),
  covariance: ArrayType(primitiveTypes.float64, 225),
});
const GetState_ResponseRefArray = ArrayType(GetState_ResponseRefStruct);
const GetState_ResponseRefStructArray = StructType({
  data: GetState_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetState_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetState_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.state = Float64Wrapper.createArray();
      this._wrapperFields.state.copy(other._wrapperFields.state);
      this._wrapperFields.covariance = Float64Wrapper.createArray();
      this._wrapperFields.covariance.copy(other._wrapperFields.covariance);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetState_ResponseRefStruct();
    this._wrapperFields.state = Float64Wrapper.createArray();
    this._wrapperFields.covariance = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetState_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetState_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetState_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetState_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetState_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    for (let i = 0; i < this._wrapperFields.state.data.length; i++) {
      this._refObject.state[i] = this._wrapperFields.state.data[i];
    }
    for (let i = 0; i < this._wrapperFields.covariance.data.length; i++) {
      this._refObject.covariance[i] = this._wrapperFields.covariance.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._refObject.state = refObject.state;
    this._wrapperFields.state.fill(refObject.state.toArray());
    this._refObject.covariance = refObject.covariance;
    this._wrapperFields.covariance.fill(refObject.covariance.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GetState_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'robot_localization', subFolder: 'srv', interfaceName: 'GetState_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get state() {
    return this._wrapperFields['state'].data;
  }
  set state(value) {
    if (value.length !== 15) {
      throw new RangeError('The length of the array must be 15.');
    }
    this._wrapperFields['state'].fill(value);
  }
  get covariance() {
    return this._wrapperFields['covariance'].data;
  }
  set covariance(value) {
    if (value.length !== 225) {
      throw new RangeError('The length of the array must be 225.');
    }
    this._wrapperFields['covariance'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetState_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.state.fill(refObject.state.toArray());
    this._wrapperFields.covariance.fill(refObject.covariance.toArray());
  }
  copy(other) {
    this._refObject = new GetState_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.state.copy(other._wrapperFields.state);
    this._wrapperFields.covariance.copy(other._wrapperFields.covariance);
  }
  static get classType() {
    return GetState_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"state","type":{"isArray":true,"arraySize":15,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"covariance","type":{"isArray":true,"arraySize":225,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"robot_localization","type":"GetState_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetState_Response"};
  }
  hasMember(name) {
    let memberNames = ["state","covariance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetState_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetState_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetState_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetState_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetState_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetState_ResponseRefArray(size);
    this._refObject = new GetState_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetState_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetState_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetState_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetState_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetState_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetState_ResponseArrayWrapper;
  }
}
module.exports = GetState_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "state",
      "type": {
        "isArray": true,
        "arraySize": 15,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "covariance",
      "type": {
        "isArray": true,
        "arraySize": 225,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "robot_localization",
    "type": "GetState_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetState_Response"
}
*/
