// rclnodejs/generated/robot_localization/robot_localization__srv__GetState_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetState_RequestRefStruct = StructType({
  time_stamp: TimeWrapper.refObjectType,
  frame_id: primitiveTypes.string,
});
const GetState_RequestRefArray = ArrayType(GetState_RequestRefStruct);
const GetState_RequestRefStructArray = StructType({
  data: GetState_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetState_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetState_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.time_stamp =  new TimeWrapper(other._wrapperFields.time_stamp);
      this._frame_idIntialized = true;
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetState_RequestRefStruct();
    this._wrapperFields.time_stamp = new TimeWrapper();
    this._frame_idIntialized = false;
    this._wrapperFields.frame_id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetState_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetState_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetState_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetState_RequestRefStructArray
  }
  static get refObjectType() {
    return GetState_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._frame_idIntialized) {
        throw new TypeError('Invalid argument: frame_id in GetState_Request');
      }
    }
    this._wrapperFields.time_stamp.freeze(own, checkConsistency);
    this._refObject.time_stamp = this._wrapperFields.time_stamp.refObject;
    if (own) {
      this._wrapperFields.frame_id.freeze(own, checkConsistency);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.time_stamp.copyRefObject(refObject.time_stamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.time_stamp);
    StringWrapper.freeStruct(refObject.frame_id);
  }
  static destoryRawROS(msg) {
    GetState_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'robot_localization', subFolder: 'srv', interfaceName: 'GetState_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get time_stamp() {
    return this._wrapperFields.time_stamp;
  }
  set time_stamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.time_stamp.copy(value);
    } else {
      this._wrapperFields.time_stamp.copy(new TimeWrapper(value));
    }
  }
  get frame_id() {
    if (!this._frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetState_RequestRefStruct(refObject.toObject());
    this._wrapperFields.time_stamp.copyRefObject(this._refObject.time_stamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
  }
  copy(other) {
    this._refObject = new GetState_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.time_stamp.copy(other._wrapperFields.time_stamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
  }
  static get classType() {
    return GetState_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"time_stamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"robot_localization","type":"GetState_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetState_Request"};
  }
  hasMember(name) {
    let memberNames = ["time_stamp","frame_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetState_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetState_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetState_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetState_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetState_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetState_RequestRefArray(size);
    this._refObject = new GetState_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetState_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetState_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetState_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetState_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetState_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetState_RequestArrayWrapper;
  }
}
module.exports = GetState_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "time_stamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "robot_localization",
    "type": "GetState_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetState_Request"
}
*/
