// rclnodejs/generated/robot_localization/robot_localization__srv__SetDatum_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GeoPoseWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeoPose.js');
const SetDatum_RequestRefStruct = StructType({
  geo_pose: GeoPoseWrapper.refObjectType,
});
const SetDatum_RequestRefArray = ArrayType(SetDatum_RequestRefStruct);
const SetDatum_RequestRefStructArray = StructType({
  data: SetDatum_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetDatum_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetDatum_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.geo_pose =  new GeoPoseWrapper(other._wrapperFields.geo_pose);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetDatum_RequestRefStruct();
    this._wrapperFields.geo_pose = new GeoPoseWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetDatum_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetDatum_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetDatum_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetDatum_RequestRefStructArray
  }
  static get refObjectType() {
    return SetDatum_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.geo_pose.freeze(own, checkConsistency);
    this._refObject.geo_pose = this._wrapperFields.geo_pose.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.geo_pose.copyRefObject(refObject.geo_pose);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GeoPoseWrapper.freeStruct(refObject.geo_pose);
  }
  static destoryRawROS(msg) {
    SetDatum_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'robot_localization', subFolder: 'srv', interfaceName: 'SetDatum_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get geo_pose() {
    return this._wrapperFields.geo_pose;
  }
  set geo_pose(value) {
    if (value instanceof GeoPoseWrapper) {
      this._wrapperFields.geo_pose.copy(value);
    } else {
      this._wrapperFields.geo_pose.copy(new GeoPoseWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetDatum_RequestRefStruct(refObject.toObject());
    this._wrapperFields.geo_pose.copyRefObject(this._refObject.geo_pose);
  }
  copy(other) {
    this._refObject = new SetDatum_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.geo_pose.copy(other._wrapperFields.geo_pose);
  }
  static get classType() {
    return SetDatum_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"geo_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeoPose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"robot_localization","type":"SetDatum_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetDatum_Request"};
  }
  hasMember(name) {
    let memberNames = ["geo_pose"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetDatum_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetDatum_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetDatum_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetDatum_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetDatum_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetDatum_RequestRefArray(size);
    this._refObject = new SetDatum_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetDatum_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetDatum_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetDatum_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetDatum_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetDatum_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetDatum_RequestArrayWrapper;
  }
}
module.exports = SetDatum_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "geo_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeoPose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "robot_localization",
    "type": "SetDatum_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetDatum_Request"
}
*/
