// rclnodejs/generated/robot_localization/robot_localization__srv__FromLL_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const FromLL_ResponseRefStruct = StructType({
  map_point: PointWrapper.refObjectType,
});
const FromLL_ResponseRefArray = ArrayType(FromLL_ResponseRefStruct);
const FromLL_ResponseRefStructArray = StructType({
  data: FromLL_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FromLL_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FromLL_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.map_point =  new PointWrapper(other._wrapperFields.map_point);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FromLL_ResponseRefStruct();
    this._wrapperFields.map_point = new PointWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FromLL_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FromLL_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return FromLL_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return FromLL_ResponseRefStructArray
  }
  static get refObjectType() {
    return FromLL_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.map_point.freeze(own, checkConsistency);
    this._refObject.map_point = this._wrapperFields.map_point.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map_point.copyRefObject(refObject.map_point);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointWrapper.freeStruct(refObject.map_point);
  }
  static destoryRawROS(msg) {
    FromLL_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'robot_localization', subFolder: 'srv', interfaceName: 'FromLL_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map_point() {
    return this._wrapperFields.map_point;
  }
  set map_point(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.map_point.copy(value);
    } else {
      this._wrapperFields.map_point.copy(new PointWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new FromLL_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.map_point.copyRefObject(this._refObject.map_point);
  }
  copy(other) {
    this._refObject = new FromLL_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.map_point.copy(other._wrapperFields.map_point);
  }
  static get classType() {
    return FromLL_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map_point","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"robot_localization","type":"FromLL_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FromLL_Response"};
  }
  hasMember(name) {
    let memberNames = ["map_point"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FromLL_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FromLL_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FromLL_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FromLL_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FromLL_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FromLL_ResponseRefArray(size);
    this._refObject = new FromLL_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FromLL_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FromLL_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FromLL_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FromLL_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FromLL_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FromLL_ResponseArrayWrapper;
  }
}
module.exports = FromLL_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "map_point",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "robot_localization",
    "type": "FromLL_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FromLL_Response"
}
*/
