// rclnodejs/generated/micro_ros_msgs/micro_ros_msgs__msg__Node.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const EntityWrapper = require('../../generated/micro_ros_msgs/micro_ros_msgs__msg__Entity.js');
const NodeRefStruct = StructType({
  node_namespace: primitiveTypes.string,
  node_name: primitiveTypes.string,
  entities: EntityWrapper.refObjectArrayType,
});
const NodeRefArray = ArrayType(NodeRefStruct);
const NodeRefStructArray = StructType({
  data: NodeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NodeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NodeRefStruct(other._refObject.toObject());
      this._node_namespaceIntialized = true;
      this._wrapperFields.node_namespace =  new StringWrapper(other._wrapperFields.node_namespace);
      this._node_nameIntialized = true;
      this._wrapperFields.node_name =  new StringWrapper(other._wrapperFields.node_name);
      this._wrapperFields.entities = EntityWrapper.createArray();
      this._wrapperFields.entities.copy(other._wrapperFields.entities);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NodeRefStruct();
    this._node_namespaceIntialized = false;
    this._wrapperFields.node_namespace = new StringWrapper();
    this._node_nameIntialized = false;
    this._wrapperFields.node_name = new StringWrapper();
    this._wrapperFields.entities = EntityWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new NodeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NodeArrayWrapper;
  }
  static get ArrayType() {
    return NodeArrayWrapper;
  }
  static get refObjectArrayType() {
    return NodeRefStructArray
  }
  static get refObjectType() {
    return NodeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._node_namespaceIntialized) {
        throw new TypeError('Invalid argument: node_namespace in Node');
      }
      if (!this._node_nameIntialized) {
        throw new TypeError('Invalid argument: node_name in Node');
      }
    }
    if (own) {
      this._wrapperFields.node_namespace.freeze(own, checkConsistency);
    }
    this._refObject.node_namespace = this._wrapperFields.node_namespace.refObject;
    if (own) {
      this._wrapperFields.node_name.freeze(own, checkConsistency);
    }
    this._refObject.node_name = this._wrapperFields.node_name.refObject;
    this._wrapperFields.entities.freeze(own, checkConsistency);
    this._refObject.entities = this._wrapperFields.entities.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._node_namespaceIntialized = true;
    this._wrapperFields.node_namespace.data = refObject.node_namespace.data;
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.data = refObject.node_name.data;
    this._wrapperFields.entities.copyRefObject(refObject.entities);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.node_namespace);
    StringWrapper.freeStruct(refObject.node_name);
    if (refObject.entities.size != 0) {
      EntityWrapper.ArrayType.freeArray(refObject.entities);
      if (EntityWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.entities, EntityWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    NodeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'micro_ros_msgs', subFolder: 'msg', interfaceName: 'Node'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get node_namespace() {
    if (!this._node_namespaceIntialized) {
      return undefined;
    }
    return this._wrapperFields.node_namespace.data;
  }
  set node_namespace(value) {
    this._node_namespaceIntialized = true;
    this._wrapperFields.node_namespace.data = value;
  }
  get node_name() {
    if (!this._node_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.node_name.data;
  }
  set node_name(value) {
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.data = value;
  }
  get entities() {
    return this._wrapperFields.entities;
  }
  set entities(value) {
    this._wrapperFields.entities.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new NodeRefStruct(refObject.toObject());
    this._node_namespaceIntialized = true;
    this._wrapperFields.node_namespace.copyRefObject(this._refObject.node_namespace);
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.copyRefObject(this._refObject.node_name);
    this._wrapperFields.entities.copyRefObject(this._refObject.entities);
  }
  copy(other) {
    this._refObject = new NodeRefStruct(other._refObject.toObject());
    this._node_namespaceIntialized = true;
    this._wrapperFields.node_namespace.copy(other._wrapperFields.node_namespace);
    this._node_nameIntialized = true;
    this._wrapperFields.node_name.copy(other._wrapperFields.node_name);
    this._wrapperFields.entities.copy(other._wrapperFields.entities);
  }
  static get classType() {
    return NodeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"node_namespace","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":256,"isPrimitiveType":true},"default_value":null},{"name":"node_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":256,"isPrimitiveType":true},"default_value":null},{"name":"entities","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"micro_ros_msgs","type":"Entity","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"micro_ros_msgs","type":"Node","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Node"};
  }
  hasMember(name) {
    let memberNames = ["node_namespace","node_name","entities"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NodeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NodeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NodeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NodeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NodeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NodeRefArray(size);
    this._refObject = new NodeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NodeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NodeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NodeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NodeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NodeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NodeArrayWrapper;
  }
}
module.exports = NodeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "node_namespace",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 256,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "node_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": 256,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "entities",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "micro_ros_msgs",
        "type": "Entity",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "micro_ros_msgs",
    "type": "Node",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Node"
}
*/
