// rclnodejs/generated/ros_gz_interfaces/ros_gz_interfaces__msg__EntityWrench.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const EntityWrapper = require('../../generated/ros_gz_interfaces/ros_gz_interfaces__msg__Entity.js');
const WrenchWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Wrench.js');
const EntityWrenchRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  entity: EntityWrapper.refObjectType,
  wrench: WrenchWrapper.refObjectType,
});
const EntityWrenchRefArray = ArrayType(EntityWrenchRefStruct);
const EntityWrenchRefStructArray = StructType({
  data: EntityWrenchRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EntityWrenchWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EntityWrenchRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.entity =  new EntityWrapper(other._wrapperFields.entity);
      this._wrapperFields.wrench =  new WrenchWrapper(other._wrapperFields.wrench);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EntityWrenchRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.entity = new EntityWrapper();
    this._wrapperFields.wrench = new WrenchWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new EntityWrenchWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EntityWrenchArrayWrapper;
  }
  static get ArrayType() {
    return EntityWrenchArrayWrapper;
  }
  static get refObjectArrayType() {
    return EntityWrenchRefStructArray
  }
  static get refObjectType() {
    return EntityWrenchRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.entity.freeze(own, checkConsistency);
    this._refObject.entity = this._wrapperFields.entity.refObject;
    this._wrapperFields.wrench.freeze(own, checkConsistency);
    this._refObject.wrench = this._wrapperFields.wrench.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.entity.copyRefObject(refObject.entity);
    this._wrapperFields.wrench.copyRefObject(refObject.wrench);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    EntityWrapper.freeStruct(refObject.entity);
    WrenchWrapper.freeStruct(refObject.wrench);
  }
  static destoryRawROS(msg) {
    EntityWrenchWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros_gz_interfaces', subFolder: 'msg', interfaceName: 'EntityWrench'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get entity() {
    return this._wrapperFields.entity;
  }
  set entity(value) {
    if (value instanceof EntityWrapper) {
      this._wrapperFields.entity.copy(value);
    } else {
      this._wrapperFields.entity.copy(new EntityWrapper(value));
    }
  }
  get wrench() {
    return this._wrapperFields.wrench;
  }
  set wrench(value) {
    if (value instanceof WrenchWrapper) {
      this._wrapperFields.wrench.copy(value);
    } else {
      this._wrapperFields.wrench.copy(new WrenchWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new EntityWrenchRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.entity.copyRefObject(this._refObject.entity);
    this._wrapperFields.wrench.copyRefObject(this._refObject.wrench);
  }
  copy(other) {
    this._refObject = new EntityWrenchRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.entity.copy(other._wrapperFields.entity);
    this._wrapperFields.wrench.copy(other._wrapperFields.wrench);
  }
  static get classType() {
    return EntityWrenchWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"entity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"ros_gz_interfaces","type":"Entity","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"wrench","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ros_gz_interfaces","type":"EntityWrench","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EntityWrench"};
  }
  hasMember(name) {
    let memberNames = ["header","entity","wrench"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EntityWrenchArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EntityWrenchWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EntityWrenchWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EntityWrenchArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EntityWrenchArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EntityWrenchRefArray(size);
    this._refObject = new EntityWrenchRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EntityWrenchWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EntityWrenchArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EntityWrenchArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EntityWrenchWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EntityWrenchWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EntityWrenchArrayWrapper;
  }
}
module.exports = EntityWrenchWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "entity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "ros_gz_interfaces",
        "type": "Entity",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "wrench",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Wrench",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros_gz_interfaces",
    "type": "EntityWrench",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EntityWrench"
}
*/
